File: sdks/js/src/result.js
//
// Copyright 2025 The Chainloop Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Creates a successful result (no violations, not skipped).
 *
 * @returns {Object} A success result
 *
 * @example
 * const result = success();
 * result.addViolation("Found an issue");
 * outputResult(result);
 */
function success() {
  return {
    skipped: false,
    violations: [],
    skip_reason: "",
    ignore: false,

    /**
     * Adds a violation message to the result.
     * @param {string} message - The violation message
     */
    addViolation(message) {
      this.violations.push(message);
    },

    /**
     * Returns true if there are any violations.
     * @returns {boolean}
     */
    hasViolations() {
      return this.violations.length > 0;
    },

    /**
     * Returns true if the policy passed (no violations, not skipped).
     * @returns {boolean}
     */
    isSuccess() {
      return !this.skipped && this.violations.length === 0;
    }
  };
}

/**
 * Creates a failed result with one or more violations.
 *
 * @param {...string} violations - One or more violation messages
 * @returns {Object} A failure result
 *
 * @example
 * const result = fail("Missing field: name", "Invalid version format");
 * outputResult(result);
 */
function fail(...violations) {
  return {
    skipped: false,
    violations: violations,
    skip_reason: "",
    ignore: false,

    addViolation(message) {
      this.violations.push(message);
    },

    hasViolations() {
      return this.violations.length > 0;
    },

    isSuccess() {
      return false;
    }
  };
}

/**
 * Creates a skipped result with a reason.
 * Use this when the policy doesn't apply to the material.
 *
 * @param {string} reason - The reason for skipping
 * @returns {Object} A skip result
 *
 * @example
 * const result = skip("Not a CycloneDX SBOM");
 * outputResult(result);
 */
function skip(reason) {
  return {
    skipped: true,
    violations: [],
    skip_reason: reason,
    ignore: false,

    addViolation(message) {
      this.violations.push(message);
    },

    hasViolations() {
      return this.violations.length > 0;
    },

    isSuccess() {
      return false;
    }
  };
}

module.exports = {
  success,
  fail,
  skip
};

File: policies/go/result.go
//
// Copyright 2025 The Chainloop Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package _go

import "fmt"

// Result represents a policy execution result.
// This type is designed to work with pdk.OutputJSON() - all fields are simple types.
type Result struct {
	Skipped    bool     `json:"skipped"`
	Violations []string `json:"violations"`
	SkipReason string   `json:"skip_reason"`
	Ignore     bool     `json:"ignore"`
}

// Success creates a successful result (no violations, not skipped).
func Success() Result {
	return Result{
		Skipped:    false,
		Violations: []string{},
		SkipReason: "",
		Ignore:     false,
	}
}

// Fail creates a failed result with one or more violations.
func Fail(violations ...string) Result {
	return Result{
		Skipped:    false,
		Violations: violations,
		SkipReason: "",
		Ignore:     false,
	}
}

// Skip creates a skipped result with a reason.
// Use this when the policy doesn't apply to the material.
func Skip(reason string) Result {
	return Result{
		Skipped:    true,
		Violations: []string{},
		SkipReason: reason,
		Ignore:     false,
	}
}

// Skipf creates a skipped result with a formatted reason.
func Skipf(format string, args ...interface{}) Result {
	return Result{
		Skipped:    true,
		Violations: []string{},
		SkipReason: fmt.Sprintf(format, args...),
		Ignore:     false,
	}
}

// AddViolation adds a violation message to the result.
func (r *Result) AddViolation(msg string) {
	r.Violations = append(r.Violations, msg)
}

// AddViolationf adds a formatted violation message to the result.
func (r *Result) AddViolationf(format string, args ...interface{}) {
	r.Violations = append(r.Violations, fmt.Sprintf(format, args...))
}

// HasViolations returns true if there are any violations.
func (r *Result) HasViolations() bool {
	return len(r.Violations) > 0
}

// IsSuccess returns true if the policy passed (no violations, not skipped).
func (r *Result) IsSuccess() bool {
	return !r.Skipped && len(r.Violations) == 0
}
